File: src/cmds/gh.ts
import { run } from '../core/runner';
import { stripAnsi, truncate } from '../core/utils';
import { HandlerResult } from './git';

// GitHub CLI. `gh` output is dense human tables; we collapse them to counts +
// the leading identifier for each row so the model can still act on them.
export function handleGh(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const verb = args[1] || '';
  const result = run('gh', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  const cmdType = `gh ${sub}${verb ? ` ${verb}` : ''}`.trim();

  try {
    if (sub === 'pr' && verb === 'list') filtered = formatPrList(result.stdout, ultra);
    else if (sub === 'pr' && verb === 'view') filtered = formatPrView(result.stdout, ultra);
    else if (sub === 'pr' && verb === 'checks') filtered = formatChecks(result.stdout, ultra);
    else if (sub === 'issue' && verb === 'list') filtered = formatIssueList(result.stdout, ultra);
    else if (sub === 'run' && verb === 'list') filtered = formatRunList(result.stdout, ultra);
    else filtered = stripAnsi(raw).trim(); // unknown gh subcommand: pass through in full
  } catch {
    filtered = raw;
  }

  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : (result.stderr || raw).trim()),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType,
    execMs: result.execMs,
  };
}

function tableRows(raw: string): string[] {
  return stripAnsi(raw)
    .split('\n')
    .map((l) => l.replace(/\r$/, ''))
    .filter((l) => l.trim() !== '' && !/^Showing\b/i.test(l));
}

function formatPrList(raw: string, ultra: boolean): string {
  const rows = tableRows(raw);
  if (rows.length === 0) return ultra ? '0 PRs' : 'no open pull requests';
  if (ultra) return `${rows.length} PRs`;
  const out = [`${rows.length} open PR${rows.length === 1 ? '' : 's'}:`];
  for (const r of rows.slice(0, 20)) {
    const cols = r.split('\t');
    const num = cols[0]?.trim() || '';
    const title = (cols[1] || '').trim().slice(0, 60);
    const branch = (cols[2] || '').trim();
    out.push(`  #${num.replace(/^#/, '')} ${title}${branch ? ` [${branch}]` : ''}`);
  }
  return out.join('\n');
}

function formatIssueList(raw: string, ultra: boolean): string {
  const rows = tableRows(raw);
  if (rows.length === 0) return ultra ? '0 issues' : 'no open issues';
  if (ultra) return `${rows.length} issues`;
  const out = [`${rows.length} open issue${rows.length === 1 ? '' : 's'}:`];
  for (const r of rows.slice(0, 20)) {
    const cols = r.split('\t');
    const num = (cols[0] || '').trim().replace(/^#/, '');
    const title = (cols[2] || cols[1] || '').trim().slice(0, 60);
    out.push(`  #${num} ${title}`);
  }
  return out.join('\n');
}

function formatRunList(raw: string, ultra: boolean): string {
  const rows = tableRows(raw);
  if (rows.length === 0) return ultra ? '0 runs' : 'no workflow runs';
  let ok = 0;
  let fail = 0;
  let other = 0;
  for (const r of rows) {
    if (/\b(completed\s+success|✓|success)\b/i.test(r)) ok++;
    else if (/\b(failure|✗|X|cancelled|timed_out)\b/i.test(r)) fail++;
    else other++;
  }
  if (ultra) return `${ok}✓/${fail}✗${other ? `/${other}?` : ''}`;
  const out = [`${rows.length} run${rows.length === 1 ? '' : 's'}: ${ok} passed, ${fail} failed${other ? `, ${other} in-progress` : ''}`];
  for (const r of rows.slice(0, 10)) {
    const cols = r.split('\t').map((c) => c.trim());
    const status = cols.find((c) => /success|failure|cancelled|in_progress|queued/i.test(c)) || cols[0] || '';
    const name = cols.find((c) => c && c !== status && !/^\d+$/.test(c)) || '';
    out.push(`  ${status} ${name}`.slice(0, 70));
  }
  return out.join('\n');
}

function formatPrView(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const title = /^title:\s*(.+)$/im.exec(clean)?.[1]?.trim() || '';
  const state = /^state:\s*(.+)$/im.exec(clean)?.[1]?.trim() || '';
  const reviewers = /^reviewers:\s*(.+)$/im.exec(clean)?.[1]?.trim() || '';
  if (ultra) return `${state || '?'}: ${title}`.slice(0, 60);
  const out = [];
  if (title) out.push(`${title}`);
  if (state) out.push(`state: ${state}`);
  if (reviewers) out.push(`reviewers: ${reviewers}`);
  // Keep the body but trimmed.
  const bodyIdx = clean.indexOf('--\n');
  if (bodyIdx >= 0) {
    const body = clean.slice(bodyIdx + 3).trim();
    if (body) out.push('', truncate(body, 15));
  }
  return out.length ? out.join('\n') : truncate(clean, 20);
}

function formatChecks(raw: string, ultra: boolean): string {
  const rows = tableRows(raw);
  let pass = 0;
  let fail = 0;
  let pending = 0;
  for (const r of rows) {
    if (/\bpass\b|✓/i.test(r)) pass++;
    else if (/\bfail\b|✗|X/i.test(r)) fail++;
    else if (/\bpending\b|in_progress/i.test(r)) pending++;
  }
  if (ultra) return `${pass}✓/${fail}✗${pending ? `/${pending}⋯` : ''}`;
  const out = [`checks: ${pass} passing, ${fail} failing${pending ? `, ${pending} pending` : ''}`];
  if (fail > 0) {
    for (const r of rows.filter((x) => /\bfail\b|✗/i.test(x)).slice(0, 10)) {
      out.push(`  ✗ ${r.split('\t')[0].trim()}`);
    }
  }
  return out.join('\n');
}

File: internal/handlers/gh.go
package handlers

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

// GitHub CLI. gh output is dense human tables; we collapse them to counts plus the
// leading identifier for each row so the model can still act on them.
func HandleGh(args []string, ultra bool) Result {
	sub, verb := "", ""
	if len(args) > 0 {
		sub = args[0]
	}
	if len(args) > 1 {
		verb = args[1]
	}
	r := run.Run("gh", args)
	raw := combined(r)
	var filtered string

	switch {
	case sub == "pr" && verb == "list":
		filtered = formatPrList(r.Stdout, ultra)
	case sub == "pr" && verb == "view":
		filtered = formatPrView(r.Stdout, ultra)
	case sub == "pr" && verb == "checks":
		filtered = formatChecks(r.Stdout, ultra)
	case sub == "issue" && verb == "list":
		filtered = formatIssueList(r.Stdout, ultra)
	case sub == "run" && verb == "list":
		filtered = formatRunList(r.Stdout, ultra)
	default:
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}

	cmdType := strings.TrimSpace("gh " + sub)
	if verb != "" {
		cmdType += " " + verb
	}
	if filtered == "" {
		if r.ExitCode == 0 {
			filtered = "ok"
		} else {
			filtered = strings.TrimSpace(firstNonEmpty(r.Stderr, raw))
		}
	}
	return Result{Filtered: filtered, Exit: r.ExitCode, Raw: raw, CmdType: cmdType, ExecMs: r.ExecMs}
}

var (
	ghShowingRe  = regexp.MustCompile(`(?i)^Showing\b`)
	ghRunOkRe    = regexp.MustCompile(`(?i)\b(completed\s+success|✓|success)\b`)
	ghRunFailRe  = regexp.MustCompile(`(?i)\b(failure|✗|X|cancelled|timed_out)\b`)
	ghRunStatRe  = regexp.MustCompile(`(?i)success|failure|cancelled|in_progress|queued`)
	ghDigitsRe   = regexp.MustCompile(`^\d+$`)
	ghPassRe     = regexp.MustCompile(`(?i)\bpass\b|✓`)
	ghFailRe     = regexp.MustCompile(`(?i)\bfail\b|✗|X`)
	ghPendingRe  = regexp.MustCompile(`(?i)\bpending\b|in_progress`)
	ghTitleRe    = regexp.MustCompile(`(?im)^title:\s*(.+)$`)
	ghStateRe    = regexp.MustCompile(`(?im)^state:\s*(.+)$`)
	ghReviewerRe = regexp.MustCompile(`(?im)^reviewers:\s*(.+)$`)
)

func tableRows(raw string) []string {
	var out []string
	for _, l := range strings.Split(util.StripAnsi(raw), "\n") {
		l = strings.TrimSuffix(l, "\r")
		if strings.TrimSpace(l) != "" && !ghShowingRe.MatchString(l) {
			out = append(out, l)
		}
	}
	return out
}

func formatPrList(raw string, ultra bool) string {
	rows := tableRows(raw)
	if len(rows) == 0 {
		if ultra {
			return "0 PRs"
		}
		return "no open pull requests"
	}
	if ultra {
		return fmt.Sprintf("%d PRs", len(rows))
	}
	out := []string{fmt.Sprintf("%d open %s:", len(rows), plural(len(rows), "PR", "PRs"))}
	for _, row := range limit(rows, 20) {
		cols := strings.Split(row, "\t")
		num := strings.TrimPrefix(strings.TrimSpace(col(cols, 0)), "#")
		title := clip(strings.TrimSpace(col(cols, 1)), 60)
		branch := strings.TrimSpace(col(cols, 2))
		line := fmt.Sprintf("  #%s %s", num, title)
		if branch != "" {
			line += " [" + branch + "]"
		}
		out = append(out, line)
	}
	return strings.Join(out, "\n")
}

func formatIssueList(raw string, ultra bool) string {
	rows := tableRows(raw)
	if len(rows) == 0 {
		if ultra {
			return "0 issues"
		}
		return "no open issues"
	}
	if ultra {
		return fmt.Sprintf("%d issues", len(rows))
	}
	out := []string{fmt.Sprintf("%d open %s:", len(rows), plural(len(rows), "issue", "issues"))}
	for _, row := range limit(rows, 20) {
		cols := strings.Split(row, "\t")
		num := strings.TrimPrefix(strings.TrimSpace(col(cols, 0)), "#")
		title := clip(strings.TrimSpace(firstNonEmpty(col(cols, 2), col(cols, 1))), 60)
		out = append(out, fmt.Sprintf("  #%s %s", num, title))
	}
	return strings.Join(out, "\n")
}

func formatRunList(raw string, ultra bool) string {
	rows := tableRows(raw)
	if len(rows) == 0 {
		if ultra {
			return "0 runs"
		}
		return "no workflow runs"
	}
	ok, fail, other := 0, 0, 0
	for _, r := range rows {
		switch {
		case ghRunOkRe.MatchString(r):
			ok++
		case ghRunFailRe.MatchString(r):
			fail++
		default:
			other++
		}
	}
	if ultra {
		s := fmt.Sprintf("%d✓/%d✗", ok, fail)
		if other != 0 {
			s += fmt.Sprintf("/%d?", other)
		}
		return s
	}
	head := fmt.Sprintf("%d %s: %d passed, %d failed", len(rows), plural(len(rows), "run", "runs"), ok, fail)
	if other != 0 {
		head += fmt.Sprintf(", %d in-progress", other)
	}
	out := []string{head}
	for _, row := range limit(rows, 10) {
		cols := strings.Split(row, "\t")
		for i, c := range cols {
			cols[i] = strings.TrimSpace(c)
		}
		status := ""
		for _, c := range cols {
			if ghRunStatRe.MatchString(c) {
				status = c
				break
			}
		}
		if status == "" {
			status = col(cols, 0)
		}
		name := ""
		for _, c := range cols {
			if c != "" && c != status && !ghDigitsRe.MatchString(c) {
				name = c
				break
			}
		}
		out = append(out, clip(fmt.Sprintf("  %s %s", status, name), 70))
	}
	return strings.Join(out, "\n")
}

func formatPrView(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	title := submatch(ghTitleRe, clean)
	state := submatch(ghStateRe, clean)
	reviewers := submatch(ghReviewerRe, clean)
	if ultra {
		return clip(fmt.Sprintf("%s: %s", firstNonEmpty(state, "?"), title), 60)
	}
	var out []string
	if title != "" {
		out = append(out, title)
	}
	if state != "" {
		out = append(out, "state: "+state)
	}
	if reviewers != "" {
		out = append(out, "reviewers: "+reviewers)
	}
	if i := strings.Index(clean, "--\n"); i >= 0 {
		body := strings.TrimSpace(clean[i+3:])
		if body != "" {
			out = append(out, "", util.Truncate(body, 15))
		}
	}
	if len(out) == 0 {
		return util.Truncate(clean, 20)
	}
	return strings.Join(out, "\n")
}

func formatChecks(raw string, ultra bool) string {
	rows := tableRows(raw)
	pass, fail, pending := 0, 0, 0
	for _, r := range rows {
		switch {
		case ghPassRe.MatchString(r):
			pass++
		case ghFailRe.MatchString(r):
			fail++
		case ghPendingRe.MatchString(r):
			pending++
		}
	}
	if ultra {
		s := fmt.Sprintf("%d✓/%d✗", pass, fail)
		if pending != 0 {
			s += fmt.Sprintf("/%d⋯", pending)
		}
		return s
	}
	head := fmt.Sprintf("checks: %d passing, %d failing", pass, fail)
	if pending != 0 {
		head += fmt.Sprintf(", %d pending", pending)
	}
	out := []string{head}
	if fail > 0 {
		var failing []string
		for _, r := range rows {
			if ghFailRe.MatchString(r) {
				failing = append(failing, r)
			}
		}
		for _, r := range limit(failing, 10) {
			out = append(out, "  ✗ "+strings.TrimSpace(strings.Split(r, "\t")[0]))
		}
	}
	return strings.Join(out, "\n")
}

func col(cols []string, i int) string {
	if i < len(cols) {
		return cols[i]
	}
	return ""
}

func submatch(re *regexp.Regexp, s string) string {
	if m := re.FindStringSubmatch(s); m != nil {
		return strings.TrimSpace(m[1])
	}
	return ""
}
